pkg/packing/packer.go

```diff
@@ -0,0 +1,151 @@
+			remainingPods = remainingPods[1:]
+			continue
+		}
+		packings = append(packings, packing)
+		zap.S().Debugf("Successfully computed packing for pod(s) %v with instance type option(s) %s", apiobject.PodNamespacedNames(packing.Pods), instanceTypeNames(packing.InstanceTypeOptions))
+	}
+	return packings
+}
+
+// packWithLargestPod will try to pack max number of pods with largest pod in
+// pods across all available node capacities. It returns Packing: max pod count
+// that fit; with their node capacities and list of leftover pods
+func (p *packer) packWithLargestPod(unpackedPods []*v1.Pod, instances []cloudprovider.InstanceType, constraints *cloudprovider.Constraints) (*Packing, []*v1.Pod) {
+	bestPackedPods := []*v1.Pod{}
+	bestInstances := []cloudprovider.InstanceType{}
+	remainingPods := unpackedPods
+	for _, packable := range PackablesFrom(instances, constraints.Overhead) {
+		// check how many pods we can fit with the available capacity
+		result := packable.Pack(unpackedPods)
+		if len(result.packed) == 0 {
+			continue
+		}
+		// If the pods packed are the same as before, this instance type can be
+		// considered as a backup option in case we get ICE
+		if p.podsMatch(bestPackedPods, result.packed) {
+			bestInstances = append(bestInstances, packable.InstanceType)
+		} else if len(result.packed) > len(bestPackedPods) {
+			// If pods packed are more than compared to what we got in last
+			// iteration, consider using this instance type
+			bestPackedPods = result.packed
+			remainingPods = result.unpacked
+			bestInstances = []cloudprovider.InstanceType{packable.InstanceType}
+		}
+	}
+	sortByResources(bestInstances)
+	return &Packing{Pods: bestPackedPods, InstanceTypeOptions: bestInstances}, remainingPods
+}
+
+func (*packer) podsMatch(first, second []*v1.Pod) bool {
+	if len(first) != len(second) {
+		return false
+	}
+	podSeen := map[string]int{}
+	for _, pod := range first {
+		podSeen[apiobject.NamespacedName(pod).String()]++
+	}
+	for _, pod := range second {
+		podSeen[apiobject.NamespacedName(pod).String()]--
+	}
+	for _, value := range podSeen {
+		if value != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// sortByResources sorts instance type packings by vcpus and memory resources
+func sortByResources(instances []cloudprovider.InstanceType) {
+	sort.Slice(instances, func(i, j int) bool { return weightOf(instances[i]) < weightOf(instances[j]) })
+}
+
+func weightOf(instance cloudprovider.InstanceType) float64 {
+	return euclidean(
+		float64(instance.CPU().Value()),
+		float64(instance.Memory().ScaledValue(resource.Mega)), // 1 gb = 1 cpu
+		float64(instance.NvidiaGPUs().Value())*1000,           // Heavily weigh gpus x 1000
+		float64(instance.AWSNeurons().Value())*1000,           // Heavily weigh neurons x1000
+	)
+}
+
+func euclidean(values ...float64) float64 {
```

Comment:
add comment explaining why "euclidean" being used?

Reply:
nit: Actually I meant more like *why* euclidean being used? This just explains what it is. (Maybe I should have mentioned the `weightOf` comment doesn't really explain why either?)

Reply:
As I'm writing up the "why", I'm remembering that we should just be using price as a single dimension ordering construct. Euclidean is just a proxy for price...